### PR TITLE
feat(rules): add tally/php/enable-opcache-in-production

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

Adds a new rule `tally/php/enable-opcache-in-production` under the PHP namespace. Flags long-running PHP web runtime final stages that don't install or enable OPcache, with two heuristic autofixes.

Implements rule #7 from [design-docs/35-php-container-rules.md §4.2](../blob/main/design-docs/35-php-container-rules.md#423-tallyphpenable-opcache-in-production).

## What triggers the rule

The final stage is treated as a PHP web runtime when any of these hold:

- base image is an official `php:*` tag containing `fpm` or `apache`
- base image is a known derivative (`serversideup/php`, `dunglas/frankenphp`, `bitnami/php-fpm`, `trafex/php-nginx`, `webdevops/php-nginx`, `webdevops/php-apache`)
- effective `ENTRYPOINT`/`CMD` is a PHP web wrapper (`php-fpm*`, `apache2-foreground`, `httpd-foreground`, `frankenphp`, `rr`)

Skips CLI-only images, dev/test/ci/debug stages, and non-final stages.

## Compliance signals (any one suppresses the violation)

- `docker-php-ext-install`/`-enable`/`-configure opcache`
- `pecl install opcache`
- apt/apk/dnf/yum install of `php*-opcache` — detected via the canonical `shell.InstallCommand` + `shell.StripPackageVersion` helpers, so `php-opcache=8.3+1ubuntu1` and `php-opcache:amd64` also match
- ini file named `*opcache*.ini`, or content (parsed with `gopkg.in/ini.v1`, so commented-out directives don't count) containing `opcache.*` keys or `zend_extension=opcache`
- `PHP_OPCACHE_*` stage env

## Autofix

Both paths emit `FixSuggestion`, `IsPreferred`, priority 88.

### 1. Official `php:*fpm*`/`*apache*` base
Insert `RUN docker-php-ext-install opcache` immediately after the final `FROM`:

```diff
 FROM php:8.4-fpm
+RUN docker-php-ext-install opcache
 WORKDIR /app
 COPY . .
```

### 2. Generic base installing PHP via a package manager
Append the matching `php*-opcache` to the existing install command, derived from the PHP package the user already chose:

| Package in `RUN` | Derived OPcache package |
|---|---|
| `php8.3-fpm` (Debian/Ubuntu) | `php8.3-opcache` |
| `php-fpm` (unversioned) | `php-opcache` |
| `php83-fpm` (Alpine) | `php83-opcache` |
| `php83-php-fpm` (Remi SCL) | `php83-php-opcache` |

```diff
 FROM debian:12-slim
-RUN apt-get install -y php8.3-fpm
+RUN apt-get install -y php8.3-fpm php8.3-opcache
 CMD ["php-fpm8.3", "-F"]
```

### No fix offered
- non-official derivatives (distro-specific remediation)
- multiple different `php*-fpm` packages (ambiguous target version)
- `CMD`/`ENTRYPOINT`-only hit with no install command in view

## Tests

- 28 unit `Check` subtests covering all gate paths and compliance signals, including ini-parser correctness (commented-out `opcache.enable=1` does not suppress)
- 7 fix-specific unit tests (official-PHP, Debian, Alpine, RHEL, derivative image no-fix, ambiguous no-fix, CMD-only no-fix)
- 1 table-driven test for the `derivePHPOpcachePackage` helper
- 1 integration lint test with snapshot
- 2 integration fix tests with snapshots (official-image path + package-manager path)

File-level coverage: **88.9%**.

## Follow-ups

- #543: refactor `no-xdebug-in-final-image` and `composer-no-dev-in-production` to use the canonical `shell.InstallCommand` helpers (they currently reimplement apt/apk/dnf/yum matching ad-hoc).

## Checklist

- [x] Rule implemented and self-registered via `init()`
- [x] Auto-fix implemented (sync, narrow zero-width insertions)
- [x] Consumes canonical `shell.InstallCommand`/`StripPackageVersion` — no ad-hoc package-manager matching
- [x] Ini content parsed semantically (not string-matched) — comments cannot cause false negatives
- [x] Unit tests cover rule behavior, fix behavior, edge cases (heredoc ini, multi-stage, dev-named stages, commented-out ini directives)
- [x] Integration lint + fix coverage with snapshots
- [x] Docs page added under `_docs/rules/tally/php/` with "What is OPcache?" educational section and the PHP manual reference
- [x] Docs navigation (`_docs/docs.json`) updated
- [x] `design-docs/PROPOSED-RULES-TRACKER.md` updated
- [x] `go test ./...`, `make lint`, `make cpd` pass

## Summary by CodeRabbit

* **New Features**
  * Added a new PHP rule to detect missing OPcache in production PHP containers and offer automated fixes for common cases.
* **Documentation**
  * Added rule documentation entry and updated the rules tracker to mark the rule as implemented.
* **Tests / Integration**
  * Added integration snapshots and new unit/integration tests validating detection and fix behavior.
* **Chores**
  * Updated an underlying dependency version.

### CODE REVIEW

File: internal/rules/tally/php/enable_opcache_in_production_test.go [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+	if len(violations) != 1 {
+		t.Fatalf("expected 1 violation, got %d", len(violations))
+	}
+	fix := violations[0].SuggestedFix
+	if fix == nil {
+		t.Fatal("expected SuggestedFix for dnf install php-fpm")
+	}
+	if fix.Edits[0].NewText != " php-opcache" {
+		t.Errorf("NewText = %q, want %q", fix.Edits[0].NewText, " php-opcache")
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production_test.go:494]: ✅ Addressed in 7c65b79b. Thanks @coderabbitai!

File: internal/rules/tally/php/enable_opcache_in_production.go [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+	runStartLine := locs[0].Start.Line // 1-based
+	editLine := runStartLine + targetPkg.Line
+	editCol := targetPkg.EndCol
+	if targetPkg.Line == 0 {
+		lineIdx := runStartLine - 1
+		if lineIdx < 0 || lineIdx >= sm.LineCount() {
+			return nil
+		}
+		editCol += shell.DockerfileRunCommandStartCol(sm.Line(lineIdx))
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go]: Thanks @coderabbitai — I looked into this but I don't think it's actually a bug. `ReconstructSourceText` strips `cmdStartCol` bytes from line 0 (`line[cmdStartCol:]` in `internal/shell/chain_format.go`). That means shell-parser column 0 on line 0 maps to Dockerfile column `cmdStartCol` — so adding `DockerfileRunCommandStartCol` is required to translate back to Dockerfile coordinates, not a double-offset. Continuation lines (`pkg.Line > 0`) are passed through unchanged by `ReconstructSourceText`, which is why the offset is guarded behind `if targetPkg.Line == 0`.

The round-trip tests I added in 7c65b79b already cover the line-0 case (Debian/Alpine/RHEL all produce correctly positioned edits). Added a continuation-line regression test in f53e4666 that exercises the `pkg.Line > 0` branch:

```dockerfile
FROM debian:12-slim
RUN apt-get install -y \
    php8.3-fpm
```

produces `RUN apt-get install -y \` / `    php8.3-fpm php8.3-opcache` as expected. Removing the `DockerfileRunCommandStartCol` addition as suggested would break the line-0 tests (edit would land `cmdStartCol` bytes too early, inside `RUN`).

File: internal/rules/tally/php/enable_opcache_in_production.go [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+	return &rules.SuggestedFix{
+		Description: "Add " + opcacheName + " to the existing " + targetIC.Manager + " " + targetIC.Subcommand,
+		Safety:      rules.FixSuggestion,
+		Priority:    priority,
+		IsPreferred: true,
+		Edits: []rules.TextEdit{{
+			Location: rules.NewRangeLocation(file, editLine, editCol, editLine, editCol),
+			NewText:  " " + opcacheName,
+		}},
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go:259]: ✅ Addressed in d86a7fe4. Thanks @coderabbitai!

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go:259]: Follow-up: tightened fixture to genuinely unsorted input `php8.3-zip php8.3-cli php8.3-fpm` (with packages both before and after `opcache` in sorted order — both rules now fire in one pass, `wantApplied: 2`); trimmed the docs interaction section to a single line and left the detailed interaction analysis in the source comment on `buildOpcachePackageManagerFix`. 4ce2829a.

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go:259]: Follow-up: switched the fix to coordinate with sort-packages via `input.IsRuleEnabled("tally/sort-packages")`. When sort-packages is enabled, the insertion anchor is the LAST package in the install command (not the `php*-fpm` token), so both fixes compose in a single `--fix` pass and produce a fully sorted list (`php8.3-cli php8.3-fpm php8.3-opcache`). Updated snapshot + docs (no more "run twice" guidance). 04621162.

File: internal/rules/tally/php/enable_opcache_in_production.go [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+			lastCmd = cc
+		}
+	}
+
+	if lastEntrypoint != nil {
+		return commandNameFromCmdLine(lastEntrypoint.CmdLine, lastEntrypoint.PrependShell)
+	}
+	if lastCmd != nil {
+		return commandNameFromCmdLine(lastCmd.CmdLine, lastCmd.PrependShell)
+	}
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go]: ✅ Addressed in 0cd8dd9b. Thanks @coderabbitai!

File: internal/rules/tally/php/enable_opcache_in_production.go [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+	if found != 1 || targetRun == nil || targetRun.Run == nil || !targetRun.UsesShell {
+		return nil
+	}
+	locs := targetRun.Run.Location()
+	if len(locs) == 0 {
+		return nil
+	}
+	runStartLine := locs[0].Start.Line // 1-based
+	editLine := runStartLine + targetPkg.Line
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go]: ✅ Addressed in e085cd62. Thanks @gemini-code-assist!

File: internal/rules/tally/php/enable_opcache_in_production.go [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+		found       int
+	)
+	for _, run := range sf.Runs {
+		if run == nil {
+			continue
+		}
+		for _, ic := range run.InstallCommands {
+			for _, pkg := range ic.Packages {
+				derived := derivePHPOpcachePackage(pkg.Normalized, ic.Manager)
+				if derived == "" {
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply internal/rules/tally/php/enable_opcache_in_production.go:199]: ✅ Addressed in ba359a37. Thanks @coderabbitai!
